lib/useOnyx.ts

```diff
@@ -0,0 +1,189 @@
+import {deepEqual} from 'fast-equals';
+import {useCallback, useEffect, useRef, useSyncExternalStore} from 'react';
+import Onyx from './Onyx';
+import type {CollectionKeyBase, OnyxCollection, OnyxKey, OnyxValue, Selector} from './types';
+import useLiveRef from './useLiveRef';
+import usePrevious from './usePrevious';
+
+type UseOnyxOptions<TKey extends OnyxKey, TReturnValue> = {
+    /**
+     * Determines if this key in this subscription is safe to be evicted.
+     */
+    canEvict?: boolean;
+
+    /**
+     * If set to false, then no data will be prefilled into the component.
+     */
+    initWithStoredValues?: boolean;
+
+    /**
+     * If set to true, data will be retrieved from cache during the first render even if there is a pending merge for the key.
+     */
+    allowStaleData?: boolean;
+
+    /**
+     * This value will be returned by the hook on the first render while the data is being read from Onyx.
+     */
+    initialValue?: OnyxValue<TKey>;
+
+    /**
+     * This will be used to subscribe to a subset of an Onyx key's data.
+     * Using this setting on `useOnyx` can have very positive performance benefits because the component will only re-render
+     * when the subset of data changes. Otherwise, any change of data on any property would normally
+     * cause the component to re-render (and that can be expensive from a performance standpoint).
+     */
+    selector?: Selector<TKey, unknown, TReturnValue>;
+};
+
+type FetchStatus = 'loading' | 'loaded';
+
+type CachedValue<TKey extends OnyxKey, TValue> = TValue extends OnyxValue<TKey> ? TValue : TKey extends CollectionKeyBase ? NonNullable<OnyxCollection<TValue>> : TValue;
+
+type ResultMetadata = {
+    status: FetchStatus;
+};
+
+type UseOnyxResult<TKey extends OnyxKey, TValue> = [CachedValue<TKey, TValue>, ResultMetadata];
+
+function getCachedValue<TKey extends OnyxKey, TValue>(key: TKey, selector?: Selector<TKey, unknown, unknown>): CachedValue<TKey, TValue> | undefined {
+    return Onyx.tryGetCachedValue(key, {selector}) as CachedValue<TKey, TValue> | undefined;
+}
+
+function useOnyx<TKey extends OnyxKey, TReturnValue = OnyxValue<TKey>>(key: TKey, options?: UseOnyxOptions<TKey, TReturnValue>): UseOnyxResult<TKey, TReturnValue> {
+    const connectionIDRef = useRef<number | null>(null);
+    const previousKey = usePrevious(key);
+
+    // Used to stabilize the selector reference and avoid unnecessary calls to `getSnapshot()`.
+    const selectorRef = useLiveRef(options?.selector);
+
+    // Stores the previous cached value as it's necessary to compare with the new value in `getSnapshot()`.
+    // We initialize it to `undefined` to simulate that we don't have any value from cache yet.
+    const cachedValueRef = useRef<CachedValue<TKey, TReturnValue> | undefined>(undefined);
```

Comment:
*question:* since this is used to store the previous cached value, why don't we use `usePrevious` for this?

Reply:
seems like you could do this:

```ts
const prevCachedValue = usePrevious(resultRef.current[0]);
```

and that might be a bit more concise

Reply:
Not sure I got all that but I'll take your word for it 😅